x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -626,14 +630,45 @@ GetUserPrivilegesResponse buildUserPrivilegesResponseObject(Role userRole) {
     }
 
     private IndexAuthorizationResult buildIndicesAccessControl(String action,
+                                                               TransportRequest request,
                                                                AuthorizationInfo authorizationInfo,
                                                                Set<String> indices,
                                                                Map<String, IndexAbstraction> aliasAndIndexLookup) {
         final Role role = ensureRBAC(authorizationInfo).getRole();
-        final IndicesAccessControl accessControl = role.authorize(action, indices, aliasAndIndexLookup, fieldPermissionsCache);
+
+        final Set<String> effectiveIndices;
+        final String targetIndex = getSingleTargetIndex(request);
+        if (targetIndex != null) {
+            effectiveIndices = indices.stream().filter(name -> {
+                if (name.equals(targetIndex)) {
+                    return true;
+                }
+                final IndexAbstraction indexAbstraction = aliasAndIndexLookup.get(name);
+                if (indexAbstraction.getType() != IndexAbstraction.Type.CONCRETE_INDEX) {
+                    return indexAbstraction.getIndices().stream().anyMatch(im -> im.getIndex().getName().equals(targetIndex));
+                }
+                return false;
+            }).collect(Collectors.toUnmodifiableSet());
+        } else {
+            effectiveIndices = indices;
+        }
+
+        final IndicesAccessControl accessControl = role.authorize(action, effectiveIndices, aliasAndIndexLookup, fieldPermissionsCache);
         return new IndexAuthorizationResult(true, accessControl);
     }
 
+    // An allowlist of requests that have only a single concrete target index regardless of the original indices
+    private String getSingleTargetIndex(TransportRequest request) {
+        if (request instanceof ShardSearchRequest) {
+            return ((ShardSearchRequest) request).shardId().getIndexName();
+        } else if (request instanceof FieldCapabilitiesIndexRequest) {
+            return ((FieldCapabilitiesIndexRequest) request).index();
+        } else if (request instanceof ShardFetchSearchRequest) {
+            return ((ShardFetchSearchRequest) request).getShardSearchRequest().shardId().getIndexName();
+        }
+        return null;
+    }
```

Comment:
There are more Transport Requests that target a single index/shard. These three are chosen to match what are currently being used in benchmarking. They also serve the purpose of demostrating the general idea of this change. I'll work on a more complete list once we agree on the approach.